Sortable headers For both Configuration Lists and Order Lists table resolves #467 and #282

end-9214 opened the pull request:
I tried to be smart and hard reset my forked repo :c  Which removed my Previous #282 PR ;c 

Anyway -

This PR includes both issues (sortable headers for configuration list table) #282  and (sortable headers for order_list tables)  #467
Closes #282 
Closes #467 

* Created clickable headers that maintain existing filters while toggling sort direction.
* Implemented ascending/descending toggle with visual indicators (↑/↓).
* Custom sorting for - configuration_list `min_media` property, Orders_list's special handling for JSON Fields.
* Maintained Sort Parameters across pagination .

Result Images -
### Configuration List table -
![Screenshot from 2025-03-09 11-24-02](https://github.com/user-attachments/assets/2d55fe4f-f522-4645-8f16-8839d8448671)
![Screenshot from 2025-03-09 11-23-03](https://github.com/user-attachments/assets/32685a66-ac2a-46b2-a2f2-bca6b9f0dc8b)


### Order List table - 

![Screenshot from 2025-03-09 10-27-38](https://github.com/user-attachments/assets/1e4e1eba-7a73-424f-b452-97c6e14d1318)
![Screenshot from 2025-03-09 10-27-21](https://github.com/user-attachments/assets/5e23bba4-dd0d-4360-add7-1e9200cf258b)
![Screenshot from 2025-03-09 10-27-05](https://github.com/user-attachments/assets/59202aef-b4b0-45b1-af8a-8344cd04a363)
![Screenshot from 2025-03-09 10-26-31](https://github.com/user-attachments/assets/a0ed5b41-3b72-447d-b644-b35fa5169eab)

end-9214 commented:
Hey @rgaudin,
If you are free, can you review this PR.

rgaudin commented on manager/manager/templatetags/manager.py:
```diff
@@ -232,3 +232,13 @@ def has_expired(errors) -> bool:
+
+@register.simple_tag
+def url_replace(request, field, value):
+    dict_ = request.GET.copy()
```
using the variable type as variable name is bad practice. Please use a name that's meaningful.

rgaudin commented on manager/manager/templatetags/manager.py:
```diff
@@ -232,3 +232,13 @@ def has_expired(errors) -> bool:
 register.filter("has_expired", has_expired)
+
+@register.simple_tag
+def url_replace(request, field, value):
```
the name of this tag is confusing. It should reflect what it does

rgaudin replied:
There's no docstring to explain what this function does. AFAIU it returns the (encoded) GET query params with the updated value for the requested field

rgaudin replied:
This is gonna be used only for sorting ; maybe we can drop the `field` param and use this function only to update sorting

rgaudin commented on manager/manager/views/config.py:
```diff
@@ -142,6 +149,49 @@ def configuration_list(request):
+    order_prefix = '' if sort_dir == 'asc' else '-'
+    
+    if sort_field == 'min_media':
+        import re
```
All imports should be at top of file (except when cycling)

rgaudin commented on manager/manager/views/config.py:
```diff
@@ -142,6 +149,49 @@ def configuration_list(request):
+    
+    if sort_field == 'min_media':
+        import re
+        configurations_list = list(queryset)
```
All this code is overly complicated, unnecessary and under-performing. Use `config.size` instead which is a model field and can thus be handled like the others

rgaudin commented on manager/manager/views/ui.py:
```diff
@@ -503,6 +503,46 @@ def address_delete(request, address_id=None):
+#order list hlper function -
+def apply_orders_sorting(queryset, sort_field, sort_dir):
+    order_prefix = '' if sort_dir == 'asc' else '-'
+    
```
Again with the trailing spaces

rgaudin commented on manager/manager/views/ui.py:
```diff
@@ -503,6 +503,46 @@ def address_delete(request, address_id=None):
+                reverse=(sort_dir == 'desc')
+            )
+            
+        preserved_order = [order.id for order in orders_list]
```
I don't understand what all this block is doing. On line 517, by sorting the queryset, you've already queried the DB for all the matching data. You already have those sorted (520/525) so there's nothing left to do I think

end-9214 replied:
i was not able to directly sort by database. So had to manually sort it in python. And due to that i had to preserve the order and use it to reorder by matching the IDs.

rgaudin commented on manager/manager/templates/configuration_list.html:
```diff
@@ -8,39 +8,43 @@ <h3>{% blocktrans with user_org=user.profile.organization %}{{ user_org }}'s con
-		<th>{% blocktrans %}Updated By{% endblocktrans %}</th>
-		<th>{% blocktrans %}Updated On{% endblocktrans %}</th>
+    <thead><tr>
+		<th><a href="?{% url_replace request 'sort' 'name' %}{% if sort_field == 'name' %}{% if sort_dir == 'asc' %}&dir=desc{% else %}&dir=asc{% endif %}{% else %}&dir=asc{% endif %}" class="text-dark">{% blocktrans %}Name{% endblocktrans %}{% if sort_field == 'name' %} {% if sort_dir == 'asc' %}↑{% else %}↓{% endif %}{% endif %}</a></th>
```
That last `&dir=asc` is making them all ASC by default

end-9214 commented:
> Thank you!
> 
> - please follow the coding style: no trailing spaces, black formatting, docstrings where it matters, ruff linting (`hatch run lint:all` should run successfully)
> - It is currently defaulting to ASC order when changing order. Given the fields that we have, I believe defaulting to DESC is better
> - Sorting Configurations by _Updated By_ raises `FieldError` : _Cannot resolve keyword 'username' into field. Choices are: can_order_physical, configurations, created_addresses, expire_on, id, language, order, organization, organization_id, passwordresets, user, user_id_
> - The sorting is applied **before** the pagination. I understand the reasoning but is this what users would expect? ATM, If I sort Config by size, I can see the smallest and the largest. When I click on next page, I've lost sorting (it's not applied to pagination links). If I click on sorting again, I do see page 2 of the overall sorted result.
> This is very confusing.  Before you fix this, @Popolechien can you express what you expect ? Should it sort what's on the screen (the 10 entries) or should it sort all entries and you navigate the results via the pagination?

Ok, waiting for popolechien's response.

Popolechien commented:
Sort all entries, not just those sshown

end-9214 commented:
@rgaudin please review the changes :D 
* set the default sort to desc
* added custom template `_sort_header.html` 
* removed trailing spaces
* used `config.size` instead of manually sorting using python in `config.py`

rgaudin reviewed (changes requested):
Thank you for your fixes ; please make sure the behavior is completely OK before requesting reviews. Reviews are time consuming.

* Table header uses two rows in certain conditions. Can be fixed easily via unbreakable space

<img width="1145" alt="Screenshot 2025-03-17 at 11 57 57" src="https://github.com/user-attachments/assets/6bb6fdd1-7dd1-4a2f-a7a5-853be2e4ac1f" />

* Apparently you're adding params to URLs and that can lead to dir not being understood/applied because the URL is incorrect

```
http://127.0.0.1:8000/orders/?sort=config_name&dir=desc&dir=desc&dir=desc&dir=asc&dir=desc&dir=asc&dir=desc&dir=asc&dir=desc&dir=asc&dir=desc&dir=desc&dir=asc&dir=desc&dir=asc&dir=desc&dir=asc&dir=desc
```

* Sort doesn't work for order config name

end-9214 commented:
> Thank you for your fixes ; please make sure the behavior is completely OK before requesting reviews. Reviews are time consuming.
> 
>     * Table header uses two rows in certain conditions. Can be fixed easily via unbreakable space
> 
> 
> <img alt="Screenshot 2025-03-17 at 11 57 57" width="1145" src="https://private-user-images.githubusercontent.com/57929/423460425-6bb6fdd1-7dd1-4a2f-a7a5-853be2e4ac1f.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NDIyOTA1MzgsIm5iZiI6MTc0MjI5MDIzOCwicGF0aCI6Ii81NzkyOS80MjM0NjA0MjUtNmJiNmZkZDEtN2RkMS00YTJmLWE3YTUtODUzYmUyZTRhYzFmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTAzMTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwMzE4VDA5MzAzOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTMwZmRiODliNDg1ODhhMDYxNWVhOTBjYjVlYjE5YmNkOGI4YzZkNGRiOGEyODRjNGNkNmJmYWNjZTcxMjk4OTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.UExD0nEdGh1lcKEVrKerZtTM6e61ShhCdKuKXKgIkoU">
> 
>     * Apparently you're adding params to URLs and that can lead to dir not being understood/applied because the URL is incorrect
> 
> 
> ```
> http://127.0.0.1:8000/orders/?sort=config_name&dir=desc&dir=desc&dir=desc&dir=asc&dir=desc&dir=asc&dir=desc&dir=asc&dir=desc&dir=asc&dir=desc&dir=desc&dir=asc&dir=desc&dir=asc&dir=desc&dir=asc&dir=desc
> ```
> 
>     * Sort doesn't work for order config name

Sure this time , i m keeping that in mind. I checked the behavior was completely OK :D 
* added `&nbsp;` for non-breaking space.
* fixed the URL parameter handling.
* sort is now correctly working for config name ( changed how it accessed config name, added string conversion and lowercase for case insensitive sorting, combined country and config name logic to reduce duplication.)

rgaudin commented:
<img width="1187" alt="Screenshot 2025-03-19 at 12 40 37" src="https://github.com/user-attachments/assets/03a23cc3-76ca-4e5d-b030-64831475afdd" />

end-9214 commented:
> <img width="1187" alt="Screenshot 2025-03-19 at 12 40 37" src="https://github.com/user-attachments/assets/03a23cc3-76ca-4e5d-b030-64831475afdd" />
> 

Should be fixed now 😶

rgaudin commented:
- go to `/configurations/`
- URL doesn't display any filtering ✅
- Ordering is Updated-by DESC and that's visible with the arrow ✅
- Click on Updated by
- URL is now `/configurations/?sort=updated_by&dir=desc` ✅
- Click again on Updated by (to invert order)
- URL is now `/configurations/?sort=updated_by&dir=desc&dir=asc` ❌

end-9214 commented:
> * go to `/configurations/`
> 
>     * URL doesn't display any filtering ✅
> 
>     * Ordering is Updated-by DESC and that's visible with the arrow ✅
> 
>     * Click on Updated by
> 
>     * URL is now `/configurations/?sort=updated_by&dir=desc` ✅
> 
>     * Click again on Updated by (to invert order)
> 
>     * URL is now `/configurations/?sort=updated_by&dir=desc&dir=asc` ❌

ooh i see, On it ! :D

end-9214 commented:
Fixed the Sorting URL -
now it will just show - `/configurations/?sort=updated_by&dir=desc` and `/configurations/?sort=updated_by&dir=asc` when the table headers are clicked

rgaudin commented:
- go to `/configurations/`
- Ordering is Updated-by DESC and that's visible with the arrow ✅
- Click on Updated by
- Ordering is still Updated-by DESC with arrow down ❌
- URL is now `/configurations/?sort=updated_by&dir=desc` ❌

end-9214 commented:
> * go to `/configurations/`
> 
>     * Ordering is Updated-by DESC and that's visible with the arrow ✅
> 
>     * Click on Updated by
> 
>     * Ordering is still Updated-by DESC with arrow down ❌
> 
>     * URL is now `/configurations/?sort=updated_by&dir=desc` ❌

hmm , i checked the bug, its happening in the `updated on` column in configurations and `created on` in orders .
Its happening due to both configurations and orders are set to `desc` by default by `updated on` and `created on` ,  when first clicked its order is changed to descending but in this case it was already descending which resulted in order being `desc` even when clicked.

Is it also happening in `updated by`??

rgaudin commented:
> Is it also happening in `updated by`??

No, sorry, I meant updated-on ; the default one.

end-9214 commented:
fixed the bug by adding a special handling for `updated on` and `created on` .

end-9214 commented:
@rgaudin and @Popolechien  , Should i also modify `all_orders.html` table to be sortable like that of `configuration_list.html` and `orders_list.html`?

rgaudin commented on manager/manager/views/ui.py:
```diff
@@ -504,6 +504,45 @@ def address_delete(request, address_id=None):
+    elif sort_field == 'created_on' or sort_field == 'status':
+        return queryset.order_by(f"{order_prefix}{sort_field}")
+    elif sort_field in ('config_name', 'country'):
+        orders_list = list(queryset)
```
You should limit the DB query here with `.only("scheduler_data")`

end-9214 replied:
done

rgaudin commented on manager/manager/views/ui.py:
```diff
@@ -504,6 +504,45 @@ def address_delete(request, address_id=None):
+                reverse=(sort_dir == 'desc')
+            )
+        if orders_list:
+            from django.db.models import Case, When, Value, IntegerField
```
This should be on top of file

rgaudin commented on manager/manager/views/ui.py:
```diff
@@ -504,6 +504,45 @@ def address_delete(request, address_id=None):
+            )
+        if orders_list:
+            from django.db.models import Case, When, Value, IntegerField
+            preserved_order = [order.id for order in orders_list]
```
There's no need to store that in a variable. It's only useful to build `preserved_order_cases`

rgaudin commented on manager/manager/views/ui.py:
```diff
@@ -504,6 +504,45 @@ def address_delete(request, address_id=None):
+            from django.db.models import Case, When, Value, IntegerField
+            preserved_order = [order.id for order in orders_list]
+            preserved_order_cases = [
+                When(id=id, then=Value(i)) 
```
Use proper variable names" `index`, `order_id`

rgaudin commented on manager/manager/views/ui.py:
```diff
@@ -504,6 +504,45 @@ def address_delete(request, address_id=None):
+                When(id=id, then=Value(i)) 
+                for i, id in enumerate(preserved_order)
+            ]
+            return Order.objects.filter(
```
You should be reusing the queryset here

end-9214 commented on manager/manager/templates/configuration_list.html:
```diff
@@ -1,46 +1,47 @@
+        <tr>
+            <th>{% sort_header request 'name' 'Name' sort_field sort_dir %}</th>
+            <th>{% sort_header request 'size' 'Size' sort_field sort_dir %}</th>
+            <th>{% sort_header request 'min_media' 'SD&nbsp;Size&nbsp;(min)' sort_field sort_dir %}</th>
```
okay so i used `style = "whitespace: nowrap;"` instead of `&nbsp`.

end-9214 commented on manager/manager/templates/_sort_header.html:
```diff
@@ -0,0 +1,7 @@
+{% load manager %}
+<a href="?{% update_query_params request 'sort' field_name %}{% if sort_field == field_name %}&dir={% if sort_dir == 'asc' %}desc{% else %}asc{% endif %}{% else %}&dir=desc{% endif %}" class="text-dark">
+    {{ display_name }}
```
got it.

end-9214 commented on manager/manager/templates/_paginator.html:
```diff
@@ -3,17 +3,17 @@
         {% if items_page.has_previous %}
-            <a href="?{{ filter_name }}={{ filter_value }}&page=1">{% blocktrans %}&laquo; first{% endblocktrans %}</a>
-            <a href="?{{ filter_name }}={{ filter_value }}&page={{ items_page.previous_page_number }}">{% blocktrans %}previous{% endblocktrans %}</a>
+            <a href="?{{ filter_name }}={{ filter_value }}&page=1{% if sort_field %}&sort={{ sort_field }}&dir={{ sort_dir }}{% endif %}">{% blocktrans %}&laquo; first{% endblocktrans %}</a>
```
sure, I created `_sorting_params.html` for this change.

end-9214 commented on manager/manager/views/config.py:
```diff
@@ -137,12 +140,25 @@ def configuration_list(request):
 
 
+# sorting function for configurations
+def apply_sorting(queryset, sort_field, sort_dir):
```
got it.

end-9214 commented on manager/manager/views/ui.py:
```diff
@@ -504,6 +504,45 @@ def address_delete(request, address_id=None):
+    elif sort_field in ('config_name', 'country'):
+        orders_list = list(queryset)
+        if sort_field == 'config_name':
+            orders_list.sort(
```
okay, I move this by adding two functions outside this helper function `get_config_name_key` and `get_recipient_country_key` .

end-9214 replied:
Okay,i created `get_sort_key` function to handle this.

rgaudin reviewed (changes requested):
We're almost there!

- Please use a non-breaking space instead of adding nowrap: ` `
- I'm getting weird URLs:

```
http://127.0.0.1:8000/configurations/?all=yes&page=2%20%20%20%20&sort=updated_on&dir=asc
http://127.0.0.1:8000/configurations/?all=yes&page=2++++&sort=size&dir=desc
```
